README.md
# milestone_progress 
[![](https://github.com/harpreetseera/MilestoneProgressView/workflows/Test%20and%20Build%20Apk/badge.svg)](https://github.com/harpreetseera/MilestoneProgressView/actions)
[![CodeCov](https://codecov.io/gh/harpreetseera/MilestoneProgressView/branch/master/graph/badge.svg)](https://codecov.io/gh/harpreetseera/MilestoneProgressView)

Flutter package for IOS and Android to display progress through milestone progress widget.


### Screenshots
<img src="https://raw.githubusercontent.com/harpreetseera/MilestoneProgressView/master/demo.gif" height="35%" width="35%"  alt="Demo gif"/>
<img src="https://raw.githubusercontent.com/harpreetseera/MilestoneProgressView/master/ss1.jpg" height="35%" width="35%"  alt="Demo screenshot"/>
<img src="https://raw.githubusercontent.com/harpreetseera/MilestoneProgressView/master/ss2.jpg" height="35%" width="35%"  alt="Demo screenshot2"/>
## Usage

[Example]https://github.com/harpreetseera/MilestoneProgressView/blob/master/example/example_app.dart)

To use this package :

* add the dependency to your [pubspec.yaml](https://github.com/harpreetseera/MilestoneProgressView/blob/master/pubspec.yaml) file.

```yaml
  dependencies:
    flutter:
      sdk: flutter
    milestone_progress:
```
## How to use

```dart
/* use this widget to create the milestone progress view
Note:The widget doesnot maintain its state itself.So to see the update , the state needs to be updated whenever a change is made .
*/
MilestoneProgress(
                completedMilestone: 3,
                maxIconSize: 30,
                totalMilestones: 6,
                width: 250,
                completedIconData: Icons.favorite, //optional 
                completedIconColor: Colors.red, //optional 
                nonCompletedIconData: Icons.favorite, //optional 
                incompleteIconColor: Colors.grey, //optional 
              )
```
# Pull Requests

I welcome and encourage all pull requests. It usually will take me within 24-48 hours to respond to any issue or request. Here are some basic rules to follow to ensure timely addition of your request:

1.  Match coding style (braces, spacing, etc.) This is best achieved using `Reformat Code` feature of Android Studio `CMD`+`Option`+`L` on Mac and `CTRL` + `ALT` + `L` on Linux + Windows .
2.  If its a feature, bugfix, or anything please only change code to what you specify.
3.  Please keep PR titles easy to read and descriptive of changes, this will make them easier to merge :)
4.  Pull requests _must_ be made against `develop` branch. Any other branch (unless specified by the maintainers) will get rejected.
5.  Check for existing [issues](https://github.com/harpreetseera/MilestoneProgressView/issues) first, before filing an issue.
6.  Make sure you follow the set standard as all other projects in this repo do
7.  Have fun!

### Created & Maintained By

[Harpreet Singh](https://github.com/harpreetseera) 

# License

    Copyright 2019 Harpreet Singh

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.



## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [0.0.4] - Release 

* Changed readme
* Added demo gif and screenshots

## [0.0.3] - Release 

* Changes in example code
* Demo gif added





test/milestone_progress_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../example/lib/main.dart';

void main() {
  group('Widget Smoke test', () {
    testWidgets('Increment Button Test', (WidgetTester tester) async {
      await tester.pumpWidget(MyApp());

      expect(find.byTooltip('Increment'), findsOneWidget);
      expect(find.byTooltip('decrement'), findsOneWidget);

      // Ensuring counter starts at 0
      expect(find.text('0 milestone completed'), findsOneWidget);

      // Incrementing counter by +1
      await tester.tap(find.byIcon(Icons.add));
      await tester.pumpAndSettle();

      // To ensure counter incremented by +1
      expect(find.text('1 milestone completed'), findsOneWidget);
    });

    testWidgets('Decrement Button Test', (WidgetTester tester) async {
      await tester.pumpWidget(MyApp());

      expect(find.byTooltip('Increment'), findsOneWidget);
      expect(find.byTooltip('decrement'), findsOneWidget);

      // Ensuring counter starts at 0
      expect(find.text('0 milestone completed'), findsOneWidget);

      // Incrementing counter by +1 to test decrement
      await tester.tap(find.byIcon(Icons.add));
      await tester.pumpAndSettle();

      // To ensure counter incremented by +1
      expect(find.text('1 milestone completed'), findsOneWidget);

      // Decrementing counter by -1
      await tester.tap(find.byIcon(Icons.remove));
      await tester.pumpAndSettle();

      // To ensure counter decremented by -1
      expect(find.text('0 milestone completed'), findsOneWidget);
    });
  });
}


example/example_app.dart
import 'package:flutter/material.dart';
import 'package:milestone_progress/milestone_progress.dart';

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'MilestoneProgressView ',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'MilestoneProgressView Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0, totalMilestones = 6;
  double maxIconSize = 30, width = 250;

  void _incrementCounter() {
    print(MediaQuery.of(context).size.width);
    if (_counter < totalMilestones)
      setState(() {
        _counter++;
      });
  }

  void _decrementCounter() {
    print(MediaQuery.of(context).size.width);
    if (_counter > 0)
      setState(() {
        _counter--;
      });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              MilestoneProgress(
                completedMilestone: _counter,
                maxIconSize: maxIconSize,
                totalMilestones: totalMilestones,
                width: width,
                completedIconData: Icons.favorite,
                completedIconColor: Colors.red,
                nonCompletedIconData: Icons.favorite,
                incompleteIconColor: Colors.grey,
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Text(
                  '$_counter milestone completed',

                ),
              ),
            ],
          ),
        ),
        floatingActionButton: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: Icon(Icons.add),
            ),
            SizedBox(
              height: 10,
            ),
            FloatingActionButton(
              onPressed: _decrementCounter,
              tooltip: 'decrement',
              child: Icon(Icons.remove),
            ),
          ],
        ));
  }
}
 

lib/milestone_progress.dart
library milestone_progress;

import 'package:flutter/material.dart';

typedef MilestoneProgressBuilder = Widget Function(
    BuildContext context, String text, Color color);

class MilestoneProgress extends StatelessWidget {
  final double width;
  final double maxIconSize;
  final int totalMilestones;
  final int completedMilestone;
  final IconData completedIconData;
  final IconData nonCompletedIconData;
  final Color completedIconColor;
  final Color incompleteIconColor;
  

  MilestoneProgress(
      {@required this.width,
      this.maxIconSize = 24,
      @required this.totalMilestones,
      @required this.completedMilestone,
      this.completedIconData = Icons.check_circle,
      this.nonCompletedIconData = Icons.adjust,
      this.completedIconColor = Colors.green,
      this.incompleteIconColor = Colors.grey})
      : assert(width != null),
        assert(totalMilestones != null),
        assert(completedMilestone != null),
        assert(totalMilestones != null);

  @override
  Widget build(BuildContext context) {
    return Row(
        mainAxisSize: MainAxisSize.min,
        children: getTitleWithMilestone(
            width,
            totalMilestones,
            completedMilestone,
            completedIconData,
            nonCompletedIconData,
            completedIconColor,
            incompleteIconColor,
            maxIconSize));
  }
}

List<Widget> getTitleWithMilestone(
    double width,
    int totalPoints,
    int checkedPoints,
    IconData completedIconData,
    IconData nonCompletedIconData,
    Color completedIconColor,
    Color incompleteIconColor,
    double mxIconSize) {
  List<Widget> list = List<Widget>();
  double iconSizeTemp = width / (1.5 * totalPoints);
  double lineSizeTemp = width / (3 * totalPoints);
  double maxIconSize = mxIconSize;
  double maxLineSize = mxIconSize / 2;

  // checks and adjusts icon size acc to max width
  double iconSize = iconSizeTemp > maxIconSize ? maxIconSize : iconSizeTemp;
  double lineSize = lineSizeTemp > maxLineSize ? maxLineSize : lineSizeTemp;

  for (int i = 0; i < (2 * totalPoints) - 1; i++) {
    if (i % 2 == 0) {
      list.add(Icon(
        i <= (2 * checkedPoints) - 1 ? completedIconData : nonCompletedIconData,
        size: iconSize,
        color: i <= (2 * checkedPoints) - 1
            ? completedIconColor
            : incompleteIconColor,
      ));
    } else {
      list.add(SizedBox(
        height: 1,
        width: lineSize,
        child: Container(
          color: i < (2 * checkedPoints) - 1
              ? completedIconColor
              : incompleteIconColor,
        ),
      ));
    }
  }
  return list;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:milestone_progress/milestone_progress.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'MilestoneProgressView ',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'MilestoneProgressView Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0, totalMilestones = 6;
  double maxIconSize = 30, width = 250;

  void _incrementCounter() {
    print(MediaQuery.of(context).size.width);
    if (_counter < totalMilestones)
      setState(() {
        _counter++;
      });
  }

  void _decrementCounter() {
    print(MediaQuery.of(context).size.width);
    if (_counter > 0)
      setState(() {
        _counter--;
      });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              MilestoneProgress(
                completedMilestone: _counter,
                maxIconSize: maxIconSize,
                totalMilestones: totalMilestones,
                width: width,
                completedIconData: Icons.favorite,
                completedIconColor: Colors.red,
                nonCompletedIconData: Icons.favorite,
                incompleteIconColor: Colors.grey,
              ),
              Padding(
                padding: const EdgeInsets.all(8.0),
                child: Text(
                  '$_counter milestone completed',

                ),
              ),
            ],
          ),
        ),
        floatingActionButton: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            FloatingActionButton(
              onPressed: _incrementCounter,
              tooltip: 'Increment',
              child: Icon(Icons.add),
            ),
            SizedBox(
              height: 10,
            ),
            FloatingActionButton(
              onPressed: _decrementCounter,
              tooltip: 'decrement',
              child: Icon(Icons.remove),
            ),
          ],
        ));
  }
}
